### PR TITLE
Add Tests for Price Type Predicates; closes #2118.

### LINKED ISSUE #2118: Add Tests for Price Type Predicates
- assets\prototype\domain\shared\predicates\priceTypes
- convert to TS
- add unit tests

### PR DESCRIPTION
**Pull request by nicolad**
## Problem this Pull Request solves
- Converts `selectionPredicates` to a folder with `index.ts` and `index.test.ts`
- Adds tests for `getDefaultPriceType`
- Closes https://github.com/eventespresso/event-espresso-core/issues/2118

### CODE REVIEW

File: assets/prototype/domain/shared/predicates/priceTypes/index.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export * from './selectionPredicates';
```
Comment (nicolad) [inline assets/prototype/domain/shared/predicates/priceTypes/index.ts]: @tn3rb should `selectionPredicates` be exposed at this level ?

**Comment by nicolad** [comment]
@tn3rb @manzoorwanijk I noticed that only `getDefaultPriceType` is being used, so I added tests only for that function.

File: assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ * ?Internal dependencies
+ */
+import { PriceBasetype } from '../../../../../domain/eventEditor/data/types';
+import { getDefaultPriceType } from './index';
+import { nodes as priceTypes } from '../../../../../domain/eventEditor/data/queries/priceTypes/test/data';
+
+describe('getDefaultPriceType', () => {
+	it('should get default price type - SURCHARGE', () => {
+		const defaultPriceType = getDefaultPriceType(priceTypes);
+		expect(defaultPriceType.baseType).toBe(PriceBasetype.SURCHARGE);
```
Comment (manzoorwanijk) [inline assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: `defaultPriceType` doesn't always need to be a `SURCHARGE`

Comment (manzoorwanijk) [reply assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: Am I right @tn3rb ?

Comment (nicolad) [reply assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: Since `getDefaultPriceType` returns `find(isFlatFeeSurcharge)(priceTypes)` I was expecting `SURCHARGE` :smiley:

Comment (tn3rb) [reply assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: Well first off, a better name for "default price type" would have been something like "default price modifier", since that is basically the default option for the price type selector in the ticket price calculator when you go to add a new price modifier.

Currently we will be using surcharge for the default price type(modifier) so what Vadim has is correct.
We could decide at some point in the future to change that, but if that happens we'll need to change the tests too.

What would be ideal is to have an option where the user could decide what they want to use for default price modifiers, then set that value in the config, and then read that value whenever we have to add a new price modifier. 

But what is here will be fine for the foreseeable future.

File: assets/prototype/domain/shared/predicates/priceTypes/index.ts (hunk 1 above)
Comment (tn3rb) [reply assets/prototype/domain/shared/predicates/priceTypes/index.ts]: i think so, the intent is for these functions to be the building blocks (ie: Lego) for more complex business logic.

**Comment by tn3rb** [comment]
> I noticed that only getDefaultPriceType is being used, so I added tests only for that function.

can you plz add tests for all of them now?

I understand that only that one function is being utilized at the moment, but as I said in another comment, these predicate functions are intended to be building blocks for more complex logic, so I think we want to verify that they operate as expected sooner rather than later. Otherwise, someone could waste time trying to use a faulty lego block in something they are trying to build and not be able to grok why things are not working. If we first confirm that these blocks all work as expected, then they can be used with confidence.

while you are at it, can you plz rename `getDefaultPriceType` to `getDefaultPriceModifierType` so that it is a little more clear as to what its purpose is.

thnx

File: assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	isBasePrice,
+	isNotBasePrice,
+	isDiscount,
+	isNotDiscount,
+	isPercent,
+	isNotPercent,
+	isTax,
+	isNotTax,
+	isFlatFeeSurcharge,
+	getDefaultPriceModifierType,
```
Comment (nicolad) [inline assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: Kept the same order in the other file.

File: assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+} from './index';
+import { nodes as priceTypes } from '../../../../../domain/eventEditor/data/queries/priceTypes/test/data';
+
+describe('isBasePrice & isNotBasePrice', () => {
+	it('should return true if priceType is base price', () => {
+		priceTypes.forEach((priceType) => {
+			if (priceType.isBasePrice === true) {
+				expect(isBasePrice(priceType)).toBe(true);
+			}
```
Comment (tn3rb) [inline assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: should maybe add an else clause to confirm that the function is not sending false positives

ie:

```js
	if (priceType.isBasePrice === true) {
		expect(isBasePrice(priceType)).toBe(true);
	} else {
		expect(isBasePrice(priceType)).toBe(false);
	}
```

or you could just compare directly against `priceType.isBasePrice`:


```js
	expect(isBasePrice(priceType)).toBe(priceType.isBasePrice);
```

maybe do the same for all of the tests using this pattern

File: assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+		});
+	});
+});
+
+describe('isFlatFeeSurcharge', () => {
+	it('should return true if price type is flat fee surcharge', () => {
+		priceTypes.forEach((priceType) => {
+			if (priceType.isBasePrice === false && priceType.isDiscount === false && priceType.isPercent === false) {
+				expect(isFlatFeeSurcharge(priceType)).toBe(true);
+			}
```
Comment (tn3rb) [inline assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: this test has at least three potential points of failure since it is testing against three properties as well as the logic that combines those into one comparison, so if any of the tests used internally by `isFlatFeeSurcharge()` fails, then it could be giving false positives, but this unit test doesn't check for that.

this one could capture the result of the conditional into a var and then compare against that

ie:

```js
	const dollarSurcharge = priceType.isBasePrice === false && priceType.isDiscount === false && priceType.isPercent === false;
	expect(isFlatFeeSurcharge(priceType)).toBe(dollarSurcharge);
```

[EDIT]

oops... just saw below that you are also testing all of the property combinations separately, so my above comment is not entirely accurate at all since you are testing other possibilities.

that said, I still don't think it's wise to apply tests conditionally like this since you are potentially missing out on an opportunity to catch some weird edge case failure.

File: assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts (hunk 4 above)
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: I like this suggestion:
`expect(isBasePrice(priceType)).toBe(priceType.isBasePrice);`
it looks a bit cleaner
thx :bowing_man:

Comment (nicolad) [reply assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: Updated: https://github.com/eventespresso/event-espresso-core/pull/2196/commits/07c28fb1b222c22f2271c489d1ed88be17491490

File: assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	});
+
+	it('should return false if it is basePrice', () => {
+		priceTypes.forEach((priceType) => {
+			const { isBasePrice, isDiscount, isPercent } = priceType;
+			const basePrice = isBasePrice === true && isDiscount === false && isPercent === false;
+
+			if (basePrice === true) {
+				expect(isFlatFeeSurcharge(priceType)).toBe(false);
+			}
```
Comment (nicolad) [inline assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: I've made these a bit explicit

Comment (nicolad) [reply assets/prototype/domain/shared/predicates/priceTypes/selectionPredicates/index.test.ts]: `if (basePrice === true)` then `.toBe(false)`
